js/pathfinding.js

```diff
@@ -5,56 +5,250 @@ import {settingsKey} from "./settings.js";
+				},
+				cacheId,
+				token
 			}
+		);
+		this.scheduleBackgroundCache();
+	}
+
+	/**
+	 * Find if any of the caches have more nodes to background cache. If there is, then schedule a background
+	 * caching job for that queue
+	 */
+	scheduleBackgroundCache() {
+		// If we already have a nextJobId, then don't start another one
+		if (this.background.nextJobId) return;
+
+		// Find the latest-used cache that has nodes left to cache
+		const latestCacheId = this.getLatestCacheId();
+		if (latestCacheId) {
+			this.background.nextJobId = window.requestIdleCallback(
+				() => this.runBackgroundCache(this.background.queues.get(latestCacheId))
+			);
+			this.scheduleTimeout();
+		}
+	}
+
+	/**
+	 * Start a timeout which, if we read the timeout time, will schedule a small amount of caching
+	 * to be performed every frame. This timeout will be cancelled every time we perform background caching.
+	 */
+	scheduleTimeout() {
+		this.cancelTimeout();
+		this.cancelAnimationFrame();
+
+		this.background.nextTimeoutId = window.setTimeout(
+			() => {
+				this.scheduleAnimationFrameCache();
+				this.background.nextTimeoutId = null;
+			},
+			Cache.backgroundCachingTimeoutMillis
+		);
+	}
+
+	/**
+	 * Schedule a small amount of caching to be done just before the next frame renders
+	 */
+	scheduleAnimationFrameCache() {
+		const latestCacheId = this.getLatestCacheId();
+		if (latestCacheId) {
+			this.background.nextAnimationFrameId = window.requestAnimationFrame(
+				() => this.runAnimationCache(this.background.queues.get(latestCacheId))
+			);
+		}
+	}
+
+	/**
+	 * Find which cache was last used and get its cache ID
+	 */
+	getLatestCacheId() {
+		return Array.from(this.lastUsed.entries())
+			.filter(entry => this.background.queues.get(entry[0]).hasNext())
+			.sort((a, b) => b[1] - a[1])[0]?.[0];
+	}
+
+	/**
+	 * Cache nodes for a short time, and then schedule another idle job to cache more nodes
+	 */
+	runBackgroundCache(queue) {
+		const endTime = performance.now() + Cache.maxBackgroundCachingMillis;
+		while (queue.hasNext() && performance.now() < endTime) {
+			this.cacheNextNode(queue);
+		}
+
+		this.background.nextJobId = null;
+		this.scheduleBackgroundCache();
+	}
+
+	/**
+	 * Cache nodes for a very short time, then schedule to cache more nodes next frame
+	 */
+	runAnimationCache(queue) {
+		const endTime = performance.now() + Cache.maxAnimationCachingMillis;
+		while (queue.hasNext() && performance.now() < endTime) {
+			this.cacheNextNode(queue);
+		}
+
+		this.background.nextAnimationFrameId = null;
+		this.scheduleAnimationFrameCache();
+	}
+
+	cacheNextNode(queue) {
+		let queueItem = queue.pop();
+		const node = getNode(queueItem.value.pos, this.nodes.get(queueItem.cacheId), queueItem.token);
+		for (let edge of node.edges) {
+			queue.push(
+				{
+					value: {
+						pos: {
+							x: edge.target.x,
+							y: edge.target.y
+						}
+					},
+					cacheId: queueItem.cacheId,
+					token: queueItem.token
```

Comment:
Are `cacheId` and token `always` the same value for every node in a single `ProcessOnceQueue`? If so, I think it would make more sense to store them in a different place, where they don't have to be repeated over and over again.

Reply:
The `cacheId` will always be the same, and with your other suggestions I'm getting around that.

We can end up caching the same cache "layer" with multiple tokens (e.g. select one token in one closed room, and then another similar token in another room). The only information we use from the token is stuff that makes up the the `cacheId`, so it would be nice to not need the token at all, but we pass it into `getSnapPointForTokenObj`.